### PR TITLE
Add support for custom webhook bodies

### PR DESCRIPTION
**Pull request by josephschorr**
This will allow users to configure the body content of outgoing webhooks for notifications

Fixes https://jira.coreos.com/browse/QUAY-1322

### CODE REVIEW

File: static/directives/create-external-notification.html [hunk 1]
```diff
@@ -130,21 +127,18 @@
                          allowed-entities="['user', 'team', 'org']"
                          ng-switch-when="entity"></div>
 
-                    <div ng-if="getHelpUrl(field, currentConfig)"
-                         style="margin-top: 10px;  margin-bottom: 10px">
-                      See: <a href="{{ getHelpUrl(field, currentConfig) }}" ng-safenewtab>{{ getHelpUrl(field, currentConfig) }}</a>
-                    </div>
+                    <!-- Template -->
```
Comment (alecmerdler) [inline static/directives/create-external-notification.html]: Can you add some screenshots of the new UI?

File: static/js/directives/ui/create-external-notification.js [hunk 2]
```diff
@@ -103,6 +103,9 @@ angular.module('quay').directive('createExternalNotification', function () {
         ApiService.createRepoNotification(data, params).then(function(resp) {
           $scope.status = '';
           $scope.notificationCreated({'notification': resp});
+        }, function(resp) {
+          $scope.status = 'error';
+          $scope.errorMessage = ApiService.getErrorMessage(resp, 'Could not create notification');
```
Comment (alecmerdler) [inline static/js/directives/ui/create-external-notification.js]: So we are only performing server-side validation of the input?

File: util/jsontemplate.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+from jsonpath_rw.lexer import JsonPathLexerError
+
+INLINE_PATH_PATTERN = r"\$\{([^}]*)\}"
+
+
+class JSONTemplateParseException(Exception):
+    """ Exception raised if a JSON template could not be parsed. """
+
+
+class JSONTemplate(object):
```
Comment (alecmerdler) [inline util/jsontemplate.py]: Advantages of this versus just using [Jinja.Template](https://jinja.palletsprojects.com/en/2.10.x/api/#jinja2.Template)?

Comment (josephschorr) [reply util/jsontemplate.py]: Simplicity and safety: This is guaranteed to produce JSON and unlike Jinja, we have no concerns about weird imports or execution of arbitrary code that could exhaust memory, or infinitely loop, etc.

File: static/js/directives/ui/create-external-notification.js (hunk 2 above)
Comment (josephschorr) [reply static/js/directives/ui/create-external-notification.js]: For the template? Yes. Doing so on the client would be a bit more work.

**Comment by dongboyan77** [comment]
Verified
Screenshot from 2019-12-26 10-21-58

**Comment by josephschorr** [comment]
@dongboyan77 Can you post a verification screenshot with a custom body as well? (unless the above is that and I missed it)

**Comment by dongboyan77** [comment]
Screenshot from 2019-12-25 17-38-32
@josephschorr

**Comment by josephschorr** [comment]
@dongboyan77 Does that mean it failed? Looks like it says the webhook was disabled

**Comment by dongboyan77** [comment]
@josephschorr It succeed, it shows disabled in screenshot because I make the network down

**Comment by SISheogorath** [comment]
Uhm, isn't the docs for this feature still missing? At least I couldn't find any, and the implemented `help_url` (https://docs.quay.io/guides/webhook-template.html) Points to a dead page.

**Pull request merged**